File: swift/Xtriever/Sources/Xtriever/XtrieverIndex.swift
import Foundation

/// A read-only, asynchronously searchable hybrid index (Feature 007).
///
/// Wraps the uniffi-generated `IndexHandle`. The Rust surface is synchronous on purpose
/// (research D2): uniffi 0.32.1 polls a Rust future on the awaiting task's thread, so a 2–3 s
/// search inside a Rust `async fn` would block Swift's cooperative pool. Instead every call runs
/// on a private **serial** dispatch queue and resumes the caller through a continuation:
///
/// - the caller's thread is never blocked (spec FR-005);
/// - calls on one instance run one at a time, in call order (FR-006). Two instances are
///   independent — each owns its Rust handle and mutex — so searches on different instances
///   can run concurrently, even over the same directory;
/// - cancelling the awaiting task is **cooperative** (FR-008): the Rust work runs to its budget,
///   then the caller gets `CancellationError` and the result is dropped; the instance stays
///   usable. The time budget (`SearchOptions.maxTimeMs`) is the bound.
///
/// `loadPath: .mmap` maps both models' weight files **and** the dense index's vectors
/// (`dense/index.bin`) read-only (ADR-0007, ADR-0009); the caller owns the precondition that no
/// other process modifies or truncates any of those files while the instance lives. The index
/// directory is opened read-only in every sense — nothing in it is modified or created — so an
/// index shipped inside the app bundle is opened **in place** (Feature 008 D11; the 007
/// lock-file caveat, report F-001, is gone).
public final class XtrieverIndex: @unchecked Sendable {
    /// Identity and configuration of the open index, read once at open.
    public let info: IndexInfo

    private let handle: IndexHandle
    private let queue = DispatchQueue(label: "dev.xtriever.index", qos: .utility)

    private init(handle: IndexHandle) {
        self.handle = handle
        self.info = handle.info()
    }

    /// Open a hybrid index read-only with the pinned embedder and, optionally, the pinned
    /// re-ranker. Model loading (~100 ms per model when mapped, more when buffered) runs off the
    /// caller's thread. The directory may be read-only (an app bundle): nothing is created in it.
    public static func open(
        indexDir: URL,
        embedderDir: URL,
        rerankerDir: URL?,
        loadPath: LoadPath = .mmap
    ) async throws -> XtrieverIndex {
        let handle: IndexHandle = try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.global(qos: .utility).async {
                continuation.resume(with: Result {
                    try IndexHandle.open(
                        indexDir: indexDir.path,
                        embedderDir: embedderDir.path,
                        rerankerDir: rerankerDir?.path,
                        loadPath: loadPath
                    )
                })
            }
        }
        // Cooperative cancellation: the open ran to completion, but a cancelled caller must not
        // receive a handle it never asked to keep.
        try Task.checkCancellation()
        return XtrieverIndex(handle: handle)
    }

    /// One search, off the caller's thread; calls on one instance run one at a time.
    ///
    /// The default options ask for 10 explained hits with the index's default depths and no
    /// budget. Set `maxTimeMs` to bound the search's own work: the dense stage is skipped or the
    /// re-ranker stops early and `stages` says so; with `strict` those become thrown errors
    /// instead. The budget is measured from the moment the Rust side is entered (FR-007);
    /// time spent queued behind an earlier search on this instance is **not** counted — a
    /// queued search still gets its full budget rather than arriving with it spent and
    /// degrading to a lexical-only result. Queue depth is the caller's to control (FR-006).
    public func search(
        _ query: String,
        options: SearchOptions = SearchOptions(k: 10, explain: true)
    ) async throws -> SearchResponse {
        let handle = self.handle
        let response: SearchResponse = try await withCheckedThrowingContinuation { continuation in
            queue.async {
                continuation.resume(with: Result { try handle.search(query: query, options: options) })
            }
        }
        // The Rust work cannot be interrupted, so cancellation is observed here, after it: the
        // response is dropped and the caller gets `CancellationError` (FR-008).
        try Task.checkCancellation()
        return response
    }
}

public extension HitExplain {
    /// The explanation under the pipeline's eight feature names, `.nan` where a stage did not
    /// see the hit — the same names and order as `xtriever_pipeline::HitExplain::features()`.
    func features() -> [(name: String, value: Float)] {
        [
            ("bm25.score", bm25Score ?? .nan),
            ("bm25.rank", bm25Rank.map(Float.init) ?? .nan),
            ("dense.score", denseScore ?? .nan),
            ("dense.rank", denseRank.map(Float.init) ?? .nan),
            ("fused.score", Float(fused)),
            ("rerank.score", rerankScore ?? .nan),
            ("rerank.rank", rerankRank.map(Float.init) ?? .nan),
            ("rerank.combined", rerankCombined.map(Float.init) ?? .nan),
        ]
    }
}

public extension Hit {
    /// The article title and the passage body — `text` split at its first blank line, the
    /// Feature 008 corpus convention (contracts/artefact.md). `nil` when the text has no blank
    /// line (an index built by someone else).
    var titleAndPassage: (title: String, passage: String)? {
        guard let range = text.range(of: "\n\n") else { return nil }
        return (String(text[..<range.lowerBound]), String(text[range.upperBound...]))
    }

    /// The Simple English Wikipedia URL of this passage's article, derived from the title line
    /// exactly as the build verified it against the snapshot (008 D6).
    var wikipediaURL: URL? {
        titleAndPassage.flatMap { Hit.wikipediaURL(forTitle: $0.title) }
    }

    /// `https://simple.wikipedia.org/wiki/` + the title with every UTF-8 byte outside
    /// `A–Z a–z 0–9 - _ . ~ /` percent-encoded (upper-case hex).
    static func wikipediaURL(forTitle title: String) -> URL? {
        var encoded = ""
        for byte in Array(title.utf8) {
            switch byte {
            case UInt8(ascii: "A")...UInt8(ascii: "Z"), UInt8(ascii: "a")...UInt8(ascii: "z"),
                 UInt8(ascii: "0")...UInt8(ascii: "9"),
                 UInt8(ascii: "-"), UInt8(ascii: "_"), UInt8(ascii: "."), UInt8(ascii: "~"), UInt8(ascii: "/"):
                encoded.append(Character(UnicodeScalar(byte)))
            default:
                encoded.append(String(format: "%%%02X", byte))
            }
        }
        return URL(string: "https://simple.wikipedia.org/wiki/" + encoded)
    }
}

public extension XtrieverError {
    /// The engine's message for this error, whichever case it is — for logs and UIs.
    var message: String {
        switch self {
        case .Schema(let message), .InvalidQuery(let message), .Corrupt(let message),
             .BudgetExhausted(let message), .Io(let message), .Backend(let message):
            return message
        case .UnknownField(let field):
            return "unknown field `\(field)`"
        case .DimensionMismatch(let expected, let actual):
            return "dimension mismatch: expected \(expected), got \(actual)"
        case .NotFound(let id):
            return "document \(id) not found"
        case .Model(let model, let message):
            return "model `\(model)`: \(message)"
        case .FingerprintMismatch(let index, let current):
            return "embedder fingerprint mismatch: index has `\(index)`, embedder is `\(current)`"
        }
    }
}

/// Where the build script stages resources inside the library bundle.
public enum HarnessResources {
    static var data: URL? { Bundle.module.url(forResource: "XtrieverData", withExtension: nil) }

    /// `XtrieverData/models/{embedder,reranker}`.
    public static var embedderDirectory: URL? { data?.appendingPathComponent("models/embedder") }
    public static var rerankerDirectory: URL? { data?.appendingPathComponent("models/reranker") }
    /// `XtrieverData/fixtures/{index,expected.json}`.
    public static var fixtureIndexDirectory: URL? { data?.appendingPathComponent("fixtures/index") }
    public static var fixtureExpected: URL? { data?.appendingPathComponent("fixtures/expected.json") }
    /// `XtrieverData/scifact/{index,queries.json,expected-scifact.json}`.
    public static var scifactIndexDirectory: URL? { data?.appendingPathComponent("scifact/index") }
    public static var scifactQueries: URL? { data?.appendingPathComponent("scifact/queries.json") }
    public static var scifactExpected: URL? { data?.appendingPathComponent("scifact/expected-scifact.json") }

    /// `XtrieverData/wikipedia/{index,queries.json,expected.json}` (Feature 008).
    public static var wikipediaIndexDirectory: URL? { data?.appendingPathComponent("wikipedia/index") }
    public static var wikipediaQueries: URL? { data?.appendingPathComponent("wikipedia/queries.json") }
    public static var wikipediaExpected: URL? { data?.appendingPathComponent("wikipedia/expected.json") }
    public static var wikipediaAttribution: URL? { data?.appendingPathComponent("wikipedia/ATTRIBUTION.txt") }

    public static var wikipediaIsBundled: Bool {
        guard let i = wikipediaIndexDirectory, let q = wikipediaQueries else { return false }
        return FileManager.default.fileExists(atPath: i.appendingPathComponent("xtriever-pipeline.json").path)
            && FileManager.default.fileExists(atPath: q.path)
    }

    public static var modelsAreBundled: Bool {
        guard let e = embedderDirectory, let r = rerankerDirectory else { return false }
        return FileManager.default.fileExists(atPath: e.appendingPathComponent("model.safetensors").path)
            && FileManager.default.fileExists(atPath: r.appendingPathComponent("model.safetensors").path)
    }

    public static var fixtureIsBundled: Bool {
        guard let i = fixtureIndexDirectory, let e = fixtureExpected else { return false }
        return FileManager.default.fileExists(atPath: i.appendingPathComponent("xtriever-pipeline.json").path)
            && FileManager.default.fileExists(atPath: e.path)
    }

    public static var scifactIsBundled: Bool {
        guard let i = scifactIndexDirectory, let q = scifactQueries else { return false }
        return FileManager.default.fileExists(atPath: i.appendingPathComponent("xtriever-pipeline.json").path)
            && FileManager.default.fileExists(atPath: q.path)
    }
}

File: android/xtriever/src/main/kotlin/dev/xtriever/android/XtrieverIndex.kt
package dev.xtriever.android

import java.io.Closeable
import java.io.File
import uniffi.xtriever_ffi.IndexHandle
import uniffi.xtriever_ffi.LoadPath

// The generated types are the surface; the module re-exports them under its own package so a
// consumer never imports `uniffi.*` directly, exactly as the Swift package re-exports them.
typealias SearchOptions = uniffi.xtriever_ffi.SearchOptions
typealias SearchResponse = uniffi.xtriever_ffi.SearchResponse
typealias Hit = uniffi.xtriever_ffi.Hit
typealias HitExplain = uniffi.xtriever_ffi.HitExplain
typealias StageReport = uniffi.xtriever_ffi.StageReport
typealias IndexInfo = uniffi.xtriever_ffi.IndexInfo
typealias RerankMode = uniffi.xtriever_ffi.RerankMode
// A type alias cannot reach a nested classifier, so the two modes are re-exported by name.
typealias RerankInterpolate = uniffi.xtriever_ffi.RerankMode.Interpolate
typealias RerankReplace = uniffi.xtriever_ffi.RerankMode.Replace
typealias XtrieverException = uniffi.xtriever_ffi.XtrieverException

/**
 * An open Xtriever index on Android (Feature 025).
 *
 * Mirrors `swift/Xtriever/Sources/Xtriever/XtrieverIndex.swift`: it opens the index, holds what
 * it opened and how long that took, and closes deterministically. It adds no capability the
 * Swift wrapper lacks (spec FR-001) and owns no retrieval logic — every number it returns is
 * the engine's.
 *
 * The directories must be real files the operating system can map. Assets inside the
 * application package are not: copy them into application storage first (research D6).
 */
class XtrieverIndex private constructor(
    private val handle: IndexHandle,
    /** Where this index and its models were opened from. */
    val paths: Paths,
    /** Wall time the open itself took, in milliseconds. */
    val openMs: Long,
    /**
     * The engine's own account of the index — documents, format version, fingerprints, depths,
     * the recorded re-rank mode and dense compaction share.
     *
     * Read once at open and held, as the Swift wrapper holds it: it describes what was opened,
     * so it stays readable after [close] instead of calling through a destroyed handle.
     */
    val info: IndexInfo,
) : Closeable {

    /** What was opened. */
    data class Paths(val indexDir: File, val embedderDir: File, val rerankerDir: File?)

    /**
     * One search, with the engine's own options record — the whole surface, nothing withheld.
     *
     * @throws XtrieverException as the engine raised it — nothing is translated or swallowed.
     */
    fun search(query: String, options: SearchOptions): SearchResponse = handle.search(query, options)

    /**
     * One search, for callers who would rather pass numbers than build a record. Every field of
     * [SearchOptions] is here: [k] hits, [depth] candidates per stage, [rerankDepth] fused
     * candidates re-scored (`0` for the fused list alone, `null` for the index's default),
     * [rerankMode] how the re-ranked head is ordered, [maxTimeMs] a time budget, [maxItems] an
     * item budget, [strict] to raise a stage's error instead of degrading, [explain] to attach
     * per-stage features.
     *
     * Negative values are refused rather than wrapped into enormous unsigned ones.
     *
     * @throws IllegalArgumentException if any count or budget is negative.
     * @throws XtrieverException as the engine raised it.
     */
    fun search(
        query: String,
        k: Int,
        depth: Int? = null,
        rerankDepth: Int? = null,
        rerankMode: RerankMode? = null,
        maxTimeMs: Long? = null,
        maxItems: Int? = null,
        strict: Boolean = false,
        explain: Boolean = false,
    ): SearchResponse = search(
        query,
        SearchOptions(
            k = k.count("k"),
            depth = depth?.count("depth"),
            rerankDepth = rerankDepth?.count("rerankDepth"),
            rerankMode = rerankMode,
            maxTimeMs = maxTimeMs?.also { require(it >= 0) { "maxTimeMs must not be negative: $it" } }?.toULong(),
            maxItems = maxItems?.count("maxItems"),
            strict = strict,
            explain = explain,
        ),
    )

    private fun Int.count(name: String): UInt {
        require(this >= 0) { "$name must not be negative: $this" }
        return toUInt()
    }

    override fun close() = handle.close()

    companion object {
        /**
         * Opens [indexDir] with the embedder at [embedderDir] and, when given, the re-ranker at
         * [rerankerDir].
         *
         * @throws IllegalStateException before anything native is touched when the processor
         *   cannot run the library (see [DeviceSupport]).
         * @throws XtrieverException when the engine refuses the directory — a missing file, a
         *   fingerprint mismatch, or a dense stage still on format version 1.
         */
        fun open(
            indexDir: File,
            embedderDir: File,
            rerankerDir: File? = null,
            mapped: Boolean = true,
        ): XtrieverIndex {
            when (val support = DeviceSupport.check()) {
                is DeviceSupport.Unsupported -> error(support.reason)
                DeviceSupport.Supported -> Unit
            }
            val started = System.nanoTime()
            val handle = IndexHandle.open(
                indexDir.absolutePath,
                embedderDir.absolutePath,
                rerankerDir?.absolutePath,
                if (mapped) LoadPath.MMAP else LoadPath.BUFFERED,
            )
            val openMs = (System.nanoTime() - started) / 1_000_000
            return XtrieverIndex(handle, Paths(indexDir, embedderDir, rerankerDir), openMs, handle.info())
        }
    }
}
